hazelcast/src/hazelcast/client/spi.cpp

```diff
@@ -2205,21 +2205,29 @@ namespace hazelcast {
                         invocation->set_event_handler(handler);
                         handler->before_listener_register();
 
-                        invocation->invoke_urgent().then([=] (boost::future<protocol::ClientMessage> f) {
-                            if (f.has_value()) {
-                                handler->on_listener_register();
-                                return;
-                            }
-                            //completes with exception, listener needs to be reregistered
-                            try_reregister_to_random_connection(connection);
-                        });
+                        std::weak_ptr<cluster_view_listener> weak_self = shared_from_this();
+                        connection::Connection *raw_conn = connection.get();
+
+                        invocation->invoke_urgent().then(
+                                [weak_self, handler, raw_conn](boost::future<protocol::ClientMessage> f) {
+                                    auto self = weak_self.lock();
+                                    if (!self)
+                                        return;
+
+                                    if (f.has_value()) {
+                                        handler->on_listener_register();
+                                        return;
+                                    }
+
+                                    //completes with exception, listener needs to be reregistered
+                                    self->try_reregister_to_random_connection(raw_conn);
```

Comment:
```suggestion
                                    self->try_reregister_to_random_connection(self.get());
```

Reply:
I don't understand. `self` is a `shared_ptr<cluster_view_listener>`, the method takes a `Connection *`.

Reply:
I meant this API: https://en.cppreference.com/w/cpp/memory/shared_ptr/get

Reply:
I see, then don't you need to pass `std::weak_ptr<Connection>` rather than raw pointer so that we do not use deleted connection in the lambda.

Reply:
It won't be dereferenced, only the address's equality to the `listener_added_connection_ ` variable is checked for synchronization purposes.

Reply:
OK, I changed the code so that it uses the connection id instead of the address. This should be safer. I've started nightly builds with the new version: [linux64](https://github.com/yemreinci/hazelcast-cpp-client/actions/runs/793671415), [linux32](https://github.com/yemreinci/hazelcast-cpp-client/actions/runs/793671823), [mac64](https://github.com/yemreinci/hazelcast-cpp-client/actions/runs/793672380). 

We can merge after everything is green.